core/trino-main/src/main/java/io/trino/sql/planner/JsonPathInvocationContext.java

```diff
@@ -0,0 +1,180 @@
+    {
+        try {
+            return invocations.get(new OperatorAndTypesKey(operatorType, leftType, rightType), () -> resolveInvocation(operatorType, leftType, rightType, session));
+        }
+        catch (ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Invocation resolveInvocation(OperatorType operatorType, Type leftType, Type rightType, Session session)
+    {
+        ResolvedFunction operator;
+        try {
+            operator = metadata.resolveOperator(session, operatorType, ImmutableList.of(leftType, rightType));
+        }
+        catch (OperatorNotFoundException e) {
+            return RESOLUTION_ERROR;
+        }
+
+        BoundSignature signature = operator.getSignature();
+
+        Optional<ResolvedFunction> leftCast = Optional.empty();
+        if (!signature.getArgumentTypes().get(0).equals(leftType) && !typeCoercion.isTypeOnlyCoercion(leftType, signature.getArgumentTypes().get(0))) {
+            try {
+                leftCast = Optional.of(metadata.getCoercion(session, leftType, signature.getArgumentTypes().get(0)));
+            }
+            catch (OperatorNotFoundException e) {
+                return RESOLUTION_ERROR;
+            }
+        }
+
+        Optional<ResolvedFunction> rightCast = Optional.empty();
+        if (!signature.getArgumentTypes().get(1).equals(rightType) && !typeCoercion.isTypeOnlyCoercion(rightType, signature.getArgumentTypes().get(1))) {
+            try {
+                rightCast = Optional.of(metadata.getCoercion(session, rightType, signature.getArgumentTypes().get(1)));
+            }
+            catch (OperatorNotFoundException e) {
+                return RESOLUTION_ERROR;
+            }
+        }
+
+        return invocation(operator, leftCast, rightCast);
+    }
+
+    private static class OperatorAndTypesKey
+    {
+        private final OperatorType operatorType;
+        private final Type leftType;
+        private final Type rightType;
+
+        public OperatorAndTypesKey(OperatorType operatorType, Type leftType, Type rightType)
+        {
+            this.operatorType = operatorType;
+            this.leftType = leftType;
+            this.rightType = rightType;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            OperatorAndTypesKey that = (OperatorAndTypesKey) o;
+            return operatorType == that.operatorType && Objects.equals(leftType, that.leftType) && Objects.equals(rightType, that.rightType);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(operatorType, leftType, rightType);
+        }
+    }
+
+    public static class Invocation
```

Comment:
maybe `ResolvedOperator` or `ResolvedCoercedOperator`?

(and change methods which return instance of this class accordingly)